File: weather-library/src/main/java/org/rcgonzalezf/weather/common/models/WeatherInfo.java
package org.rcgonzalezf.weather.common.models;

import android.os.Parcel;

public final class WeatherInfo implements WeatherViewModel {

  public WeatherInfo() {
  }

  private int mWeatherId;
  private int mCityId;
  private String mCityName;
  private double mSpeed;
  private double mTemp;
  private String mHumidity;
  private String mDateTime;
  private String mCountry;
  private double mDeg;
  private String mDescription;

  @Override public int describeContents() {
    return 0;
  }

  @Override public void writeToParcel(Parcel dest, int flags) {
    dest.writeInt(getId());
    dest.writeString(getCityName());
    dest.writeDouble(getSpeed());
    dest.writeDouble(getTemperature());
    dest.writeString(getHumidity());
    dest.writeString(getDateTime());
    dest.writeInt(getWeatherId());
    dest.writeString(getCountry());
    dest.writeDouble(getDeg());
    dest.writeString(getDescription());
  }

  public static final Creator<WeatherInfo> CREATOR = new Creator<WeatherInfo>() {
    public WeatherInfo createFromParcel(Parcel in) {
      return new WeatherInfo(in);
    }

    public WeatherInfo[] newArray(int size) {
      return new WeatherInfo[size];
    }
  };

  private WeatherInfo(Parcel in) {
    mCityId = in.readInt();
    mCityName = in.readString();
    mSpeed = in.readDouble();
    mTemp = in.readDouble();
    mHumidity = in.readString();
    mDateTime = in.readString();
    mWeatherId = in.readInt();
    setCountry(in.readString());
    mDeg = in.readDouble();
    mDescription = in.readString();
  }

  @Override public int getId() {
    return mCityId;
  }

  public void setCityId(int cityId) {
    this.mCityId = cityId;
  }

  @Override public String getCityName() {
    return mCityName;
  }

  public void setCityName(String cityName) {
    this.mCityName = cityName;
  }

  @Override public double getSpeed() {
    return mSpeed;
  }

  public void setSpeed(double speed) {
    this.mSpeed = speed;
  }

  @Override public double getTemperature() {
    return mTemp;
  }

  public void setTemperature(double temp) {
    this.mTemp = temp;
  }

  @Override public String getHumidity() {
    return mHumidity;
  }

  public void setHumidity(String humitidy) {
    this.mHumidity = humitidy;
  }

  @Override public String getDateTime() {
    return mDateTime;
  }

  @Override public int getWeatherId() {
    return mWeatherId;
  }

  public void setDateTime(String dateTime) {
    this.mDateTime = dateTime;
  }

  public void setWeatherId(int weatherId) {
    mWeatherId = weatherId;
  }

  @Override public String getCountry() {
    return mCountry;
  }

  @Override public double getDeg() {
    return mDeg;
  }

  @Override public String getDescription() {
    return mDescription;
  }

  public void setCountry(String mCountry) {
    this.mCountry = mCountry;
  }

  public void setDeg(double deg) {
    mDeg = deg;
  }

  public void setDescription(String description) {
    this.mDescription = description;
  }

  @Override public boolean equals(Object other) {
    if (this == other) return true;
    if (other == null || getClass() != other.getClass()) return false;

    WeatherInfo weatherInfo = (WeatherInfo) other;

    if (mWeatherId != weatherInfo.mWeatherId) return false;
    if (mCityId != weatherInfo.mCityId) return false;
    if (Double.compare(weatherInfo.mSpeed, mSpeed) != 0) return false;
    if (Double.compare(weatherInfo.mTemp, mTemp) != 0) return false;
    if (Double.compare(weatherInfo.mDeg, mDeg) != 0) return false;
    if (!mCityName.equals(weatherInfo.mCityName)) return false;
    if (mHumidity != null ? !mHumidity.equals(weatherInfo.mHumidity) : weatherInfo.mHumidity != null) {
      return false;
    }
    if (mDateTime != null ? !mDateTime.equals(weatherInfo.mDateTime) : weatherInfo.mDateTime != null) {
      return false;
    }
    if (!mCountry.equals(weatherInfo.mCountry)) return false;
    return mDescription != null ? mDescription.equals(weatherInfo.mDescription)
        : weatherInfo.mDescription == null;
  }

  @Override public int hashCode() {
    int result;
    long temp;
    result = mWeatherId;
    result = 31 * result + mCityId;
    result = 31 * result + mCityName.hashCode();
    temp = Double.doubleToLongBits(mSpeed);
    result = 31 * result + (int) (temp ^ (temp >>> 32));
    temp = Double.doubleToLongBits(mTemp);
    result = 31 * result + (int) (temp ^ (temp >>> 32));
    result = 31 * result + (mHumidity != null ? mHumidity.hashCode() : 0);
    result = 31 * result + (mDateTime != null ? mDateTime.hashCode() : 0);
    result = 31 * result + mCountry.hashCode();
    temp = Double.doubleToLongBits(mDeg);
    result = 31 * result + (int) (temp ^ (temp >>> 32));
    result = 31 * result + (mDescription != null ? mDescription.hashCode() : 0);
    return result;
  }
}

File: weather-interface/src/main/java/org/rcgonzalezf/weather/common/models/WeatherInfo.kt
package org.rcgonzalezf.weather.common.models

import android.os.Parcel
import android.os.Parcelable

class WeatherInfo : WeatherViewModel {
    constructor()

    override var weatherId = 0
    override var id = 0
        private set
    override var cityName: String? = null
    override var speed = 0.0
    override var temperature = 0.0
    override var humidity: String? = null
    override var dateTime: String? = null
    override var country: String? = null
    override var deg = 0.0
    override var description: String? = null
    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeString(cityName)
        dest.writeDouble(speed)
        dest.writeDouble(temperature)
        dest.writeString(humidity)
        dest.writeString(dateTime)
        dest.writeInt(weatherId)
        dest.writeString(country)
        dest.writeDouble(deg)
        dest.writeString(description)
    }

    private constructor(`in`: Parcel) {
        id = `in`.readInt()
        cityName = `in`.readString()
        speed = `in`.readDouble()
        temperature = `in`.readDouble()
        humidity = `in`.readString()
        dateTime = `in`.readString()
        weatherId = `in`.readInt()
        country = `in`.readString()
        deg = `in`.readDouble()
        description = `in`.readString()
    }

    fun setCityId(cityId: Int) {
        id = cityId
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val weatherInfo = other as WeatherInfo
        if (weatherId != weatherInfo.weatherId) return false
        if (id != weatherInfo.id) return false
        if (java.lang.Double.compare(weatherInfo.speed, speed) != 0) return false
        if (java.lang.Double.compare(weatherInfo.temperature, temperature) != 0) return false
        if (java.lang.Double.compare(weatherInfo.deg, deg) != 0) return false
        if (cityName != weatherInfo.cityName) return false
        if (if (humidity != null) humidity != weatherInfo.humidity else weatherInfo.humidity != null) {
            return false
        }
        if (if (dateTime != null) dateTime != weatherInfo.dateTime else weatherInfo.dateTime != null) {
            return false
        }
        if (country != weatherInfo.country) return false
        return if (description != null) description == weatherInfo.description else weatherInfo.description == null
    }

    override fun hashCode(): Int {
        var result: Int
        var temp: Long
        result = weatherId
        result = 31 * result + id
        result = 31 * result + cityName.hashCode()
        temp = java.lang.Double.doubleToLongBits(speed)
        result = 31 * result + (temp xor (temp ushr 32)).toInt()
        temp = java.lang.Double.doubleToLongBits(temperature)
        result = 31 * result + (temp xor (temp ushr 32)).toInt()
        result = 31 * result + if (humidity != null) humidity.hashCode() else 0
        result = 31 * result + if (dateTime != null) dateTime.hashCode() else 0
        result = 31 * result + country.hashCode()
        temp = java.lang.Double.doubleToLongBits(deg)
        result = 31 * result + (temp xor (temp ushr 32)).toInt()
        result = 31 * result + if (description != null) description.hashCode() else 0
        return result
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<WeatherInfo> = object : Parcelable.Creator<WeatherInfo> {
            override fun createFromParcel(`in`: Parcel): WeatherInfo? {
                return WeatherInfo(`in`)
            }

            override fun newArray(size: Int): Array<WeatherInfo?> {
                return arrayOfNulls(size)
            }
        }
    }
}